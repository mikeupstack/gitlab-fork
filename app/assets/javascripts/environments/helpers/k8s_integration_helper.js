import { differenceInSeconds } from '~/lib/utils/datetime_utility';

export function generateServicePortsString(ports) {
  if (!ports?.length) return '';

  return ports
    .map((port) => {
      const nodePort = port.nodePort ? `:${port.nodePort}` : '';
      return `${port.port}${nodePort}/${port.protocol}`;
    })
    .join(', ');
}

export function getServiceAge(creationTimestamp) {
  if (!creationTimestamp) return '';

  const timeDifference = differenceInSeconds(new Date(creationTimestamp), new Date());

  const seconds = Math.floor(timeDifference);
  const minutes = Math.floor(seconds / 60) % 60;
  const hours = Math.floor(seconds / 60 / 60) % 24;
  const days = Math.floor(seconds / 60 / 60 / 24);

  let ageString;
  if (days > 0) {
    ageString = `${days}d`;
  } else if (hours > 0) {
    ageString = `${hours}h`;
  } else if (minutes > 0) {
    ageString = `${minutes}m`;
  } else {
    ageString = `${seconds}s`;
  }

  return ageString;
}
