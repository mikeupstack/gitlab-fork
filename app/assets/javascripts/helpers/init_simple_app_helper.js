import Vue from 'vue';

/**
 * Initializes a component as a simple vue app, passing the necessary props. If the element
 * has a data attribute named `data-view-model`, the content of that attributed will be
 * converted from json and passed on to the component as a prop. The root component is then
 * responsible for setting up it's children, injections, and other desired features.
 *
 * @param {string} selector css selector for where to build
 * @param {Vue.component} component The Vue compoment to be built as the root of the app
 *
 * @example
 * ```html
 * <div id='#mount-here' data-view-model="{'some': 'object'}" />
 * ```
 *
 * ```javascript
 * initSimpleApp('#mount-here', MyApp)
 * ```
 *
 * This will mount MyApp as root on '#mount-here'. It will receive {'some': 'object'} as it's
 * view model prop.
 */
export const initSimpleApp = (selector, component) => {
  const element = document.querySelector(selector);

  if (!element) {
    return null;
  }

  const props = element.dataset.viewModel ? JSON.parse(element.dataset.viewModel) : {};

  return new Vue({
    el: element,
    render(h) {
      return h(component, { props });
    },
  });
};
