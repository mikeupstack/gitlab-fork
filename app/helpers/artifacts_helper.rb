# frozen_string_literal: true

module ArtifactsHelper
  def artifacts_app_data(project)
    {
      project_path: project.full_path,
      project_id: project.id,
      can_destroy_artifacts: can?(current_user, :destroy_artifacts, project).to_s,
      artifacts_management_feedback_image_path: image_path('illustrations/chat-bubble-sm.svg')
    }
  end
end
