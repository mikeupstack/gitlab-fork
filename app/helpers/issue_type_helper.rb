# frozen_string_literal: true

module IssueTypeHelper
  def issue_type_for(issue)
    return if issue.blank?

    if Feature.enabled?(:issue_type_uses_work_item_types_table)
      issue.work_item_type.base_type
    else
      issue.issue_type
    end
  end
end
