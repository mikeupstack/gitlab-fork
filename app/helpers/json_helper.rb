# frozen_string_literal: true

module JsonHelper
  # These two JSON helpers are short-form wrappers for the Gitlab::Json
  # class, which should be used in place of .to_json calls or calls to
  # the JSON class.
  def json_generate(...)
    Gitlab::Json.generate(...)
  end

  def json_parse(...)
    Gitlab::Json.parse(...)
  end
end
