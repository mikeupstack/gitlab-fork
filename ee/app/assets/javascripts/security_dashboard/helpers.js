import isPlainObject from 'lodash/isPlainObject';
import { REPORT_TYPES_ALL } from 'ee/security_dashboard/store/constants';
import { convertObjectPropsToSnakeCase } from '~/lib/utils/common_utils';
import { __ } from '~/locale';

/**
 * Provided a security reports summary from the GraphQL API, this returns an array of arrays
 * representing a properly formatted report ready to be displayed in the UI. Each sub-array consists
 * of the user-friend report's name, and the summary's payload. Note that summary entries are
 * considered empty and are filtered out of the return if the payload is `null` or don't include
 * a vulnerabilitiesCount property. Report types whose name can't be matched to a user-friendly
 * name are filtered out as well.
 *
 * Take the following summary for example:
 * {
 *   containerScanning: { vulnerabilitiesCount: 123 },
 *   invalidReportType: { vulnerabilitiesCount: 123 },
 *   dast: null,
 * }
 *
 * The formatted summary would look like this:
 * [
 *   ['containerScanning', { vulnerabilitiesCount: 123 }]
 * ]
 *
 * Note that `invalidReportType` was filtered out as it can't be matched with a user-friendly name,
 * and the DAST report was omitted because it's empty (`null`).
 *
 * @param {Object} rawSummary
 * @returns {Array}
 */
export const getFormattedSummary = (rawSummary = {}) => {
  if (!isPlainObject(rawSummary)) {
    return [];
  }
  // Convert keys to snake case so they can be matched against REPORT_TYPES keys for translation
  const snakeCasedSummary = convertObjectPropsToSnakeCase(rawSummary);
  // Convert object to an array of entries to make it easier to loop through
  const summaryEntries = Object.entries(snakeCasedSummary);
  // Filter out empty entries as we don't want to display those in the summary
  const withoutEmptyEntries = summaryEntries.filter(
    ([, scanSummary]) => scanSummary?.vulnerabilitiesCount !== undefined,
  );
  // Replace keys with translations found in REPORT_TYPES if available
  const formattedEntries = withoutEmptyEntries.map(([scanType, scanSummary]) => {
    const name = REPORT_TYPES_ALL[scanType];
    return name ? [name, scanSummary] : null;
  });
  // Filter out keys that could not be matched with any translation and are thus considered invalid
  return formattedEntries.filter((entry) => entry !== null);
};

export const PROJECT_LOADING_ERROR_MESSAGE = __('An error occurred while retrieving projects.');

export default () => ({});
