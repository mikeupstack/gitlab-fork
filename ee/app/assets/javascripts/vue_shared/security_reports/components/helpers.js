import { EMPTY_BODY_MESSAGE } from './constants';

/**
 * A helper function which validates the passed
 * in body string.
 *
 * It returns an empty string if the body has explicitly
 * been passed in as an empty string, a fallback
 * message if the body is null / undefined, else
 * it will return the original body string.
 *
 * @param {String} body the body message
 *
 * @return {String} the validated body message
 */
export const bodyWithFallBack = (body) => (body === '' ? '' : body || EMPTY_BODY_MESSAGE);

/**
 * Returns a string representation of the HTTP request
 *
 * @param {Object} httpData
 * @param {String} httpData.body
 * @param {String} httpData.method
 * @param {String} httpData.url
 * @param {{name: String, value: String}[]} httpData.headers
 *
 * @returns {String}
 */
export const getHttpString = (httpData) => {
  if (!httpData) {
    return '';
  }

  const { body, headers, method, url, statusCode, reasonPhrase } = httpData;
  const methodOrStatusCode = method || statusCode;
  const urlOrReasonPhrase = url || reasonPhrase;
  const headerString = headers.map(({ name, value }) => `${name}: ${value}`).join('\n');

  return `${methodOrStatusCode} ${urlOrReasonPhrase}\n${headerString}\n\n${bodyWithFallBack(body)}`;
};

export const getCreatedIssueForVulnerability = (vulnerability) =>
  vulnerability.issue_links?.find((link) => link.link_type === 'created');

export const getDismissalTransitionForVulnerability = (vulnerability) => {
  const latestTransition = vulnerability.state_transitions?.at(-1);
  return latestTransition?.to_state.toLowerCase() === 'dismissed' ? latestTransition : null;
};
