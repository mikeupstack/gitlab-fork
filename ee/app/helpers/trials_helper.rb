# frozen_string_literal: true

module TrialsHelper
  TRIAL_ONBOARDING_SOURCE_URLS = %w[about.gitlab.com docs.gitlab.com learn.gitlab.com].freeze

  def create_lead_form_data
    {
      submit_path: create_lead_trials_path(glm_params),
      first_name: current_user.first_name,
      last_name: current_user.last_name,
      company_name: current_user.organization
    }.merge(
      params.slice(
        :first_name, :last_name, :company_name, :company_size, :phone_number, :country, :state
      ).to_unsafe_h.symbolize_keys
    )
  end

  def create_company_form_data
    submit_params = glm_params.merge(passed_through_params.to_unsafe_h.symbolize_keys)
    {
      submit_path: users_sign_up_company_path(submit_params)
    }
  end

  def should_ask_company_question?
    TRIAL_ONBOARDING_SOURCE_URLS.exclude?(glm_params[:glm_source])
  end

  def glm_params
    strong_memoize(:glm_params) do
      params.slice(:glm_source, :glm_content).to_unsafe_h
    end
  end

  def glm_source
    ::Gitlab.config.gitlab.host
  end

  def trial_selection_intro_text
    if any_trialable_group_namespaces?
      s_('Trials|You can apply your trial to a new group or an existing group.')
    else
      s_('Trials|Create a new group to start your GitLab Ultimate trial.')
    end
  end

  def show_trial_namespace_select?
    any_trialable_group_namespaces?
  end

  def namespace_options_for_listbox
    group_options = trialable_group_namespaces.map { |n| { text: n.name, value: n.id.to_s } }
    options = [
      {
        text: _('New'),
        options: [
          {
            text: _('Create group'),
            value: '0'
          }
        ]
      }
    ]

    options.push(text: _('Groups'), options: group_options) unless group_options.empty?

    options
  end

  def only_trialable_group_namespace
    trialable_group_namespaces.first if trialable_group_namespaces.count == 1
  end

  private

  def passed_through_params
    params.slice(
      :trial,
      :role,
      :registration_objective,
      :jobs_to_be_done_other
    )
  end

  def trialable_group_namespaces
    strong_memoize(:trialable_group_namespaces) do
      current_user.manageable_groups_eligible_for_trial
    end
  end

  def any_trialable_group_namespaces?
    trialable_group_namespaces.any?
  end
end
