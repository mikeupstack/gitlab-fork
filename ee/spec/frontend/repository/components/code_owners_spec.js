import { shallowMount, mount } from '@vue/test-utils';
import Vue, { nextTick } from 'vue';
import VueApollo from 'vue-apollo';
import createMockApollo from 'helpers/mock_apollo_helper';
import waitForPromises from 'helpers/wait_for_promises';
import CodeOwners from 'ee_component/vue_shared/components/code_owners/code_owners.vue';
import codeOwnersInfoQuery from 'ee/graphql_shared/queries/code_owners_info.query.graphql';
import { extendedWrapper } from 'helpers/vue_test_utils_helper';
import {
  codeOwnersPath,
  codeOwnerMock,
  codeOwnersMultipleMock,
  codeOwnersPropsMock,
} from '../mock_data';

let wrapper;
let mockResolver;

const createComponent = async ({
  mountFn = shallowMount,
  props = {},
  codeOwnerDataMock = [codeOwnerMock],
} = {}) => {
  Vue.use(VueApollo);

  const project = {
    id: '1234',
    repository: {
      codeOwnersPath,
      blobs: {
        nodes: [{ id: '345', codeOwners: codeOwnerDataMock }],
      },
    },
  };

  mockResolver = jest.fn().mockResolvedValue({ data: { project } });

  wrapper = extendedWrapper(
    mountFn(CodeOwners, {
      apolloProvider: createMockApollo([[codeOwnersInfoQuery, mockResolver]]),
      propsData: {
        ...codeOwnersPropsMock,
        ...props,
      },
    }),
  );

  await waitForPromises();
};

describe('Code owners component', () => {
  const findHelpIcon = () => wrapper.findByTestId('help-icon');
  const findUsersIcon = () => wrapper.findByTestId('users-icon');
  const findCodeOwners = () => wrapper.findAllByTestId('code-owners');
  const findCommaSeparators = () => wrapper.findAllByTestId('comma-separator');
  const findAndSeparators = () => wrapper.findAllByTestId('and-separator');
  const findToggle = () => wrapper.findByTestId('collapse-toggle');
  const findBranchRulesLink = () => wrapper.findByTestId('branch-rules-link');
  const findLinkToFile = () => wrapper.findByTestId('codeowners-file-link');
  const findLinkToDocs = () => wrapper.findByTestId('codeowners-docs-link');

  beforeEach(() => createComponent());

  describe('help link', () => {
    it('renders a link to CODEOWNERS file', () => {
      expect(findLinkToFile().attributes('href')).toBe(codeOwnersPath);
    });

    it('renders a link to docs component', () => {
      expect(findLinkToDocs().attributes('href')).toBe('/help/user/project/code_owners');
      expect(findLinkToDocs().attributes('target')).toBe('_blank');
      expect(findLinkToDocs().attributes('title')).toBe('About this feature');
    });

    it('renders a Help icon', () => {
      expect(findHelpIcon().exists()).toBe(true);
      expect(findHelpIcon().props('name')).toBe('question-o');
    });
  });

  it('renders a Users icon', () => {
    expect(findUsersIcon().exists()).toBe(true);
    expect(findUsersIcon().props('name')).toBe('users');
  });

  it('doesn`t render toggle when the number of codeowners is less than 5', () => {
    expect(findToggle().exists()).toBe(false);
  });

  it('does not render a link to branch rules settings for non-maintainers', async () => {
    await createComponent({ props: { canViewBranchRules: false } });
    expect(findBranchRulesLink().exists()).toBe(false);
  });

  it('renders a link to branch rules settings for users with maintainer access and higher', () => {
    expect(findBranchRulesLink().attributes('href')).toBe(codeOwnersPropsMock.branchRulesPath);
  });

  it.each`
    codeOwners                            | commaSeparators | codeOwnersLength
    ${[]}                                 | ${0}            | ${0}
    ${[codeOwnerMock]}                    | ${0}            | ${1}
    ${codeOwnersMultipleMock.slice(0, 2)} | ${1}            | ${2}
    ${codeOwnersMultipleMock.slice(0, 3)} | ${2}            | ${3}
    ${codeOwnersMultipleMock}             | ${4}            | ${5}
  `('matches the snapshot', async ({ codeOwners, commaSeparators, codeOwnersLength }) => {
    await createComponent({ codeOwnerDataMock: codeOwners });

    expect(findCommaSeparators().length).toBe(commaSeparators);
    expect(findCodeOwners().length).toBe(codeOwnersLength);
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('when the number of code owners is more than 5', () => {
    beforeEach(() =>
      createComponent({ codeOwnerDataMock: codeOwnersMultipleMock, mountFn: mount }),
    );

    it('renders a toggle button with correct text', () => {
      expect(findToggle().exists()).toBe(true);
      expect(findToggle().text()).toBe('2 more');
    });

    it('renders only first 5 codeowners at initial state', () => {
      expect(findCodeOwners().length).toBe(5);
    });

    it('expands when you click on a toggle', async () => {
      await findToggle().trigger('click');
      await nextTick();
      expect(findCodeOwners().length).toBe(codeOwnersMultipleMock.length);
      expect(findToggle().text()).toBe('show less');
      expect(findAndSeparators().length).toBe(1);
    });
  });
});
