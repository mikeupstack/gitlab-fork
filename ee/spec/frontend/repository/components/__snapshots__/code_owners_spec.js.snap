// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code owners component matches the snapshot 1`] = `<!---->`;

exports[`Code owners component matches the snapshot 2`] = `
<div
  class="well-segment blob-auxiliary-viewer file-owner-content gl-display-flex gl-justify-content-space-between gl-align-items-center"
>
  <div
    class="gl-display-flex gl-flex-wrap"
  >
    <div
      class="gl-display-inline gl-mr-2"
    >
      <gl-icon-stub
        data-testid="users-icon"
        name="users"
        size="16"
      />
       
      <gl-link-stub
        class="gl-font-weight-bold gl-text-black-normal!"
        data-testid="codeowners-file-link"
        href="path/to/codeowners/file"
      >
        Code owners
      </gl-link-stub>
       
      <gl-link-stub
        data-testid="codeowners-docs-link"
        href="/help/user/project/code_owners"
        target="_blank"
        title="About this feature"
      >
        <gl-icon-stub
          data-testid="help-icon"
          name="question-o"
          size="16"
        />
      </gl-link-stub>
      
      :
    
    </div>
     
    <div
      data-testid="code-owners"
    >
      <!---->
       
      <gl-link-stub
        href="path/to/@johnDoe"
        target="_blank"
      >
        
        John Doe
      
      </gl-link-stub>
    </div>
     
    <!---->
     
    <!---->
  </div>
   
  <gl-button-stub
    buttontextclasses=""
    category="primary"
    class="gl-ml-4"
    data-testid="branch-rules-link"
    href="/some/Project/-/settings/repository#js-branch-rules"
    icon=""
    size="small"
    variant="default"
  >
    
    Manage branch rules
  
  </gl-button-stub>
</div>
`;

exports[`Code owners component matches the snapshot 3`] = `
<div
  class="well-segment blob-auxiliary-viewer file-owner-content gl-display-flex gl-justify-content-space-between gl-align-items-center"
>
  <div
    class="gl-display-flex gl-flex-wrap"
  >
    <div
      class="gl-display-inline gl-mr-2"
    >
      <gl-icon-stub
        data-testid="users-icon"
        name="users"
        size="16"
      />
       
      <gl-link-stub
        class="gl-font-weight-bold gl-text-black-normal!"
        data-testid="codeowners-file-link"
        href="path/to/codeowners/file"
      >
        Code owners
      </gl-link-stub>
       
      <gl-link-stub
        data-testid="codeowners-docs-link"
        href="/help/user/project/code_owners"
        target="_blank"
        title="About this feature"
      >
        <gl-icon-stub
          data-testid="help-icon"
          name="question-o"
          size="16"
        />
      </gl-link-stub>
      
      :
    
    </div>
     
    <div
      data-testid="code-owners"
    >
      <!---->
       
      <gl-link-stub
        href="/raisa"
        target="_blank"
      >
        
        Idella Welch
      
      </gl-link-stub>
    </div>
    <div
      data-testid="code-owners"
    >
      <span
        data-testid="comma-separator"
      >
        , 
      </span>
       
      <gl-link-stub
        href="/noella"
        target="_blank"
      >
        
        Winston Von
      
      </gl-link-stub>
    </div>
     
    <!---->
     
    <!---->
  </div>
   
  <gl-button-stub
    buttontextclasses=""
    category="primary"
    class="gl-ml-4"
    data-testid="branch-rules-link"
    href="/some/Project/-/settings/repository#js-branch-rules"
    icon=""
    size="small"
    variant="default"
  >
    
    Manage branch rules
  
  </gl-button-stub>
</div>
`;

exports[`Code owners component matches the snapshot 4`] = `
<div
  class="well-segment blob-auxiliary-viewer file-owner-content gl-display-flex gl-justify-content-space-between gl-align-items-center"
>
  <div
    class="gl-display-flex gl-flex-wrap"
  >
    <div
      class="gl-display-inline gl-mr-2"
    >
      <gl-icon-stub
        data-testid="users-icon"
        name="users"
        size="16"
      />
       
      <gl-link-stub
        class="gl-font-weight-bold gl-text-black-normal!"
        data-testid="codeowners-file-link"
        href="path/to/codeowners/file"
      >
        Code owners
      </gl-link-stub>
       
      <gl-link-stub
        data-testid="codeowners-docs-link"
        href="/help/user/project/code_owners"
        target="_blank"
        title="About this feature"
      >
        <gl-icon-stub
          data-testid="help-icon"
          name="question-o"
          size="16"
        />
      </gl-link-stub>
      
      :
    
    </div>
     
    <div
      data-testid="code-owners"
    >
      <!---->
       
      <gl-link-stub
        href="/raisa"
        target="_blank"
      >
        
        Idella Welch
      
      </gl-link-stub>
    </div>
    <div
      data-testid="code-owners"
    >
      <span
        data-testid="comma-separator"
      >
        , 
      </span>
       
      <gl-link-stub
        href="/noella"
        target="_blank"
      >
        
        Winston Von
      
      </gl-link-stub>
    </div>
    <div
      data-testid="code-owners"
    >
      <span
        data-testid="comma-separator"
      >
        , 
      </span>
       
      <gl-link-stub
        href="/edyth"
        target="_blank"
      >
        
        Don Runte
      
      </gl-link-stub>
    </div>
     
    <!---->
     
    <!---->
  </div>
   
  <gl-button-stub
    buttontextclasses=""
    category="primary"
    class="gl-ml-4"
    data-testid="branch-rules-link"
    href="/some/Project/-/settings/repository#js-branch-rules"
    icon=""
    size="small"
    variant="default"
  >
    
    Manage branch rules
  
  </gl-button-stub>
</div>
`;

exports[`Code owners component matches the snapshot 5`] = `
<div
  class="well-segment blob-auxiliary-viewer file-owner-content gl-display-flex gl-justify-content-space-between gl-align-items-center"
>
  <div
    class="gl-display-flex gl-flex-wrap"
  >
    <div
      class="gl-display-inline gl-mr-2"
    >
      <gl-icon-stub
        data-testid="users-icon"
        name="users"
        size="16"
      />
       
      <gl-link-stub
        class="gl-font-weight-bold gl-text-black-normal!"
        data-testid="codeowners-file-link"
        href="path/to/codeowners/file"
      >
        Code owners
      </gl-link-stub>
       
      <gl-link-stub
        data-testid="codeowners-docs-link"
        href="/help/user/project/code_owners"
        target="_blank"
        title="About this feature"
      >
        <gl-icon-stub
          data-testid="help-icon"
          name="question-o"
          size="16"
        />
      </gl-link-stub>
      
      :
    
    </div>
     
    <div
      data-testid="code-owners"
    >
      <!---->
       
      <gl-link-stub
        href="/raisa"
        target="_blank"
      >
        
        Idella Welch
      
      </gl-link-stub>
    </div>
    <div
      data-testid="code-owners"
    >
      <span
        data-testid="comma-separator"
      >
        , 
      </span>
       
      <gl-link-stub
        href="/noella"
        target="_blank"
      >
        
        Winston Von
      
      </gl-link-stub>
    </div>
    <div
      data-testid="code-owners"
    >
      <span
        data-testid="comma-separator"
      >
        , 
      </span>
       
      <gl-link-stub
        href="/edyth"
        target="_blank"
      >
        
        Don Runte
      
      </gl-link-stub>
    </div>
    <div
      data-testid="code-owners"
    >
      <span
        data-testid="comma-separator"
      >
        , 
      </span>
       
      <gl-link-stub
        href="/julietta_rogahn"
        target="_blank"
      >
        
        Sherri McClure
      
      </gl-link-stub>
    </div>
    <div
      data-testid="code-owners"
    >
      <span
        data-testid="comma-separator"
      >
        , 
      </span>
       
      <gl-link-stub
        href="/youlanda"
        target="_blank"
      >
        
        Amber Koch
      
      </gl-link-stub>
    </div>
     
    <!---->
     
    <span
      class="gl-ml-2"
    >
       and 
    </span>
     
    <gl-button-stub
      buttontextclasses=""
      category="primary"
      class="gl-ml-2"
      data-testid="collapse-toggle"
      icon=""
      size="medium"
      variant="link"
    >
      
        2 more
      
    </gl-button-stub>
  </div>
   
  <gl-button-stub
    buttontextclasses=""
    category="primary"
    class="gl-ml-4"
    data-testid="branch-rules-link"
    href="/some/Project/-/settings/repository#js-branch-rules"
    icon=""
    size="small"
    variant="default"
  >
    
    Manage branch rules
  
  </gl-button-stub>
</div>
`;
